src/app/SimpleSubscriptionResumptionStorage.h

```diff
@@ -0,0 +1,136 @@
+ *
+ *    Licensed under the Apache License, Version 2.0 (the "License");
+ *    you may not use this file except in compliance with the License.
+ *    You may obtain a copy of the License at
+ *
+ *        http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *    Unless required by applicable law or agreed to in writing, software
+ *    distributed under the License is distributed on an "AS IS" BASIS,
+ *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *    See the License for the specific language governing permissions and
+ *    limitations under the License.
+ */
+
+/**
+ *    @file
+ *      This file defines a basic implementation of SubscriptionResumptionStorage that
+ *      persists subscriptions in a flat list in TLV.
+ */
+
+#pragma once
+
+#include <app/SubscriptionResumptionStorage.h>
+#include <lib/core/TLV.h>
+#include <lib/support/DefaultStorageKeyAllocator.h>
+#include <lib/support/Pool.h>
+
+namespace chip {
+namespace app {
+
+/**
+ * An example SubscriptionResumptionStorage using PersistentStorageDelegate as it backend.
+ */
+class SimpleSubscriptionResumptionStorage : public SubscriptionResumptionStorage
+{
+public:
+    static constexpr size_t kIteratorsMax = CHIP_CONFIG_MAX_GROUP_CONCURRENT_ITERATORS;
+
+    CHIP_ERROR Init(PersistentStorageDelegate * storage);
+
+    SubscriptionInfoIterator * IterateSubscriptions() override;
+
+    CHIP_ERROR Save(SubscriptionInfo & subscriptionInfo) override;
+
+    CHIP_ERROR Delete(NodeId nodeId, FabricIndex fabricIndex, SubscriptionId subscriptionId) override;
+
+    CHIP_ERROR DeleteAll(FabricIndex fabricIndex) override;
+
+protected:
+    CHIP_ERROR Save(TLV::TLVWriter & writer, SubscriptionInfo & subscriptionInfo);
+    CHIP_ERROR Load(uint16_t subscriptionIndex, SubscriptionInfo & subscriptionInfo);
+    CHIP_ERROR Delete(uint16_t subscriptionIndex);
+    uint16_t Count();
+    CHIP_ERROR DeleteMaxCount();
+
+    class SimpleSubscriptionInfoIterator : public SubscriptionInfoIterator
+    {
+    public:
+        SimpleSubscriptionInfoIterator(SimpleSubscriptionResumptionStorage & storage);
+        size_t Count() override;
+        bool Next(SubscriptionInfo & output) override;
+        void Release() override;
+
+    private:
+        SimpleSubscriptionResumptionStorage & mStorage;
+        uint16_t mNextIndex;
+    };
+
+    static constexpr size_t MaxScopedNodeIdSize() { return TLV::EstimateStructOverhead(sizeof(NodeId), sizeof(FabricIndex)); }
+
+    static constexpr size_t MaxSubscriptionPathsSize()
+    {
+        // IM engine declares an attribute path pool and an event path pool, and each pool
+        // includes CHIP_IM_SERVER_MAX_NUM_PATH_GROUPS_FOR_SUBSCRIPTIONS for subscriptions
+        return 2 *
+            TLV::EstimateStructOverhead(
+                   TLV::EstimateStructOverhead(sizeof(uint8_t), sizeof(EndpointId), sizeof(ClusterId), sizeof(AttributeId)) *
+                   CHIP_IM_SERVER_MAX_NUM_PATH_GROUPS_FOR_SUBSCRIPTIONS);
```

Comment:
I don't think this is correct:

* assumes `sizeof(EventId) == sizeof(AttributeId)`, which is true, but needs an assert or comment
* neglects the `EventPathType` flag for events
* multiplies only `AttributeId` / `EventId` by `CHIP_IM_SERVER_MAX_NUM_PATH_GROUPS_FOR_SUBSCRIPTIONS`, when all of endpoint+cluster+attribute/event[+urgent] should be multiplied by this

Also, the TLV structure for the PR is encoding `pathCount` literals, and then the array items.  But we have a formal TLV Array container type to do this.

This is another reason to attempt to reuse the TLV path structures IM  uses on the wire for persistence if possible.

Reply:
Hmm, I'm pretty sure the inner most bit is `TLV::EstimateStructOverhead()` of `sizeof(uint8_t), sizeof(EndpointId), sizeof(ClusterId), sizeof(AttributeId)`, and that all together is multiplied by `CHIP_IM_SERVER_MAX_NUM_PATH_GROUPS_FOR_SUBSCRIPTIONS`. The uint8_t bit is to account for the `EventPathType`. And so the estimate should be more inclusive than necessary, perhaps. I could calculate attributes separate from events and likely get a lower estimate.

As mentioned earlier, I've filed #24535 for myself and will try to move toward more code reuse before 1.1

Reply:
Apologies, I misread.  But it's an odd setup because (I think) the outer `TLV::EstimateStructOverhead` call is effectively giving the space for `pathCount` with the 4-byte overhead, instead of just calling out the needed space for `pathCount` explicitly.

I'm concerned at deferring #24535 though.  The current TLV structure that's being written doesn't seem legal:

> A.5.1. Structures
A structure is a collection of member elements that each have a distinct meaning. All member
elements within a structure SHALL have a unique tag as compared to the other members of the
structure.

We are repeating the array elements in the top level of the structure with repeated tags and no array container.  Even if this works with our code today, it could well break if our future TLV code enforces the spec more strictly.

Reply:
I see. Let me rectify that.

Reply:
I've put the attributes and events into Lists of Structures.

Reply:
Thank you.